### ISSUE TITLE
Memory-efficient attention

### ISSUE THREAD

**Issue by justheuristic** [learning-at-home/lean_transformer#7]
This is a discussion of how to minimize memory usage of attention.

**Comment by justheuristic** [comment]
Summary based on @krunt 's recent talk about FMHA design:

- there are two ways to implement attention:
  -- the naive way:
       - 1. compute query-key dot-products by the tile, store dot-products in global memory, 
       - 2. then compute attention weights: softmax of dot-products, store results in global memory
       - 3. then compute the weighted sum of values with attention weights store sums in shared memory
  -- the shmemory way:
        - load a subset of queries and all keys/values from global to shared memory
        - compute dot-products, maintain them in shared memory without offloading to global
        - compute attention weights via softmax in-place in shared memory
        - then compute weighted sum of values with attention weights **and only then store results in global memory**


The shmemory way is significantly faster (~10x on fmha benchmark #8), but requires that all keys/values fit into shared memory. As a result, both FMHA and FasterTransformer are limited by head dimension 64 and sequence length 512.

In turn, the naive way supports arbitrary head size and sequence length, but is significantly slower because it needs to store/load intermediate values in global memory.

**Comment by justheuristic** [comment]
Based on these two solutions, we can produce a middle-of-the-road implementation that the flexibility of naive strategy with __most_of_the__ performance from shmemory-based strategy

### Stage 1: compute log-sum-exps
for each query, compute a scalar log-sum-exp of dot products, i.e. 
`result[i] = log(sum_over_j(<query_i, key_j>))`

Log-sum-exps can be partially computed in chunks of `tile_size` tokens.
Second, third, etc. tiles do the following:

```python
# forall tile i = 0...num_queries/tile_size, j=0...num_keys/tile_size
logaddexp_accumulators_i = load_logsumexp_outputs_from_previous_part()  # initially 1d[tile_size] of -inf
new_log_add_exps_ij = compute_dotproduct_logsumexp(query_tiles[i], key_tiles[j])
logaddexp_accumulators_i [:]= safe_logaddexp_pair(logaddexp_accumulators_i, new_log_add_exps_ij)
```

Wherein `compute_dotproduct_logsumexp` stands for computing dot-product of queries to keys, followed by a [reduce_logsumexp](https://pytorch.org/docs/stable/generated/torch.logsumexp.html) over all keys, parallel for each query.
, safe_logaddexp_pair is an element-wise log-sum of two exponents, equivalent to [torch logaddexp](https://pytorch.org/docs/stable/generated/torch.logaddexp.html)

__i/o:__ `load queries and keys, 2x [tile_size x head size], store logsumexps: small [tile_size] vectors`
__flops:__ ~half of fusedMHA's forward pass, since we have no need

### Stage 2: forward (given logsumexp)

Once we know log-sum-exps, we no longer need to load the entire set of queries into shared memory.

Instead, we can load one chunk at a time, compute partial attention outputs from that chunk, add them to the accumulator, then load the next chunk, etc.

```python
# forall tile i = 0...num_queries/tile_size, j=0...num_keys/tile_size
query_tiles[i], key_tiles[j], value_tiles[j] = load_into_shmemory()
attention_accumulators_i = load_partial_results_from_previous_part()  # initially 2d[num_queries, head_dim] of zeros
logsumexp_accumulator_i = load_from_stage_1_for_queries_i()

dot_product_ij = dot_product(query_tiles[i], key_tiles[j])
softmax_tile_ij = exp(dot_product_ij - logsumexp_accumulator_i)
attention_output_tile_ij = dot_product(softmax_tile_ij, value_tiles[j])
attention_accumulators_i [:]= attention_accumulators_i + attention_output_tile_ij

```

### Stage 3: backward

Use the same backward logic as in shmemory, but this time you reuse log-sum-exps saved from the forward pass and accumulate gradients by tiles.


### Notes:
- compute log-add-exps during forward pass and reuse for backward pass
- if it works well enough, maybe contribute this to apex in order to avoid compilation here?

**Comment by krunt** [comment]
![image](https://user-images.githubusercontent.com/5300844/163032215-aa66e83e-2b06-4630-8600-21d9fc36fd95.png)

fwd fmha for longer sequences is implented on this fork https://github.com/krunt/apex

**Comment by krunt** [comment]
bwd is implemented:

![image](https://user-images.githubusercontent.com/5300844/163872565-13cdd09d-35d9-4009-9444-0af586b19b91.png)

**Comment by krunt** [comment]
fwd+bwd results:

![image](https://user-images.githubusercontent.com/5300844/163872731-88f192e6-4de3-41a9-8d82-246eacb633f2.png)

![image](https://user-images.githubusercontent.com/5300844/163872745-e47365df-6a85-4b4c-a932-521fb4aa0f8c.png)

![image](https://user-images.githubusercontent.com/5300844/163872762-0559d74f-c694-4a55-bdde-aea42d6df66c.png)
